File: src/main/java/com/atsuishio/superbwarfare/client/overlay/IFFOverlay.java
package com.atsuishio.superbwarfare.client.overlay;

import com.atsuishio.superbwarfare.Mod;
import com.atsuishio.superbwarfare.config.client.DisplayConfig;
import com.atsuishio.superbwarfare.entity.projectile.MineEntity;
import com.atsuishio.superbwarfare.entity.vehicle.base.VehicleEntity;
import com.atsuishio.superbwarfare.init.ModItems;
import com.atsuishio.superbwarfare.tools.SeekTool;
import com.atsuishio.superbwarfare.tools.VectorTool;
import com.atsuishio.superbwarfare.tools.VectorUtil;
import com.mojang.blaze3d.platform.GlStateManager;
import com.mojang.blaze3d.systems.RenderSystem;
import net.minecraft.client.Camera;
import net.minecraft.client.DeltaTracker;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.LayeredDraw;
import net.minecraft.client.renderer.GameRenderer;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.Mth;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.entity.vehicle.Boat;
import net.minecraft.world.level.ClipContext;
import net.minecraft.world.phys.HitResult;
import net.minecraft.world.phys.Vec3;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.api.distmarker.OnlyIn;
import top.theillusivec4.curios.api.CuriosApi;

import javax.annotation.ParametersAreNonnullByDefault;
import java.util.List;

import static com.atsuishio.superbwarfare.client.RenderHelper.preciseBlit;

@OnlyIn(Dist.CLIENT)
public class IFFOverlay implements LayeredDraw.Layer {

    public static final ResourceLocation ID = Mod.loc("iff");

    public static final ResourceLocation FRIENDLY_INDICATOR = Mod.loc("textures/screens/teammate/friendly_indicator.png");
    public static final ResourceLocation FRIENDLY_AIRCRAFT = Mod.loc("textures/screens/teammate/friendly_aircraft.png");
    public static final ResourceLocation FRIENDLY_TANK = Mod.loc("textures/screens/teammate/friendly_tank.png");
    public static final ResourceLocation FRIENDLY_APC = Mod.loc("textures/screens/teammate/friendly_apc.png");
    public static final ResourceLocation FRIENDLY_AA = Mod.loc("textures/screens/teammate/friendly_aa.png");
    public static final ResourceLocation FRIENDLY_CAR = Mod.loc("textures/screens/teammate/friendly_car.png");
    public static final ResourceLocation FRIENDLY_ARTILLERY = Mod.loc("textures/screens/teammate/friendly_artillery.png");
    public static final ResourceLocation FRIENDLY_BOAT = Mod.loc("textures/screens/teammate/friendly_boat.png");
    public static final ResourceLocation FRIENDLY_DEFENSE = Mod.loc("textures/screens/teammate/friendly_defense.png");
    public static final ResourceLocation FRIENDLY_DRONE = Mod.loc("textures/screens/teammate/friendly_drone.png");
    public static final ResourceLocation FRIENDLY_HELICOPTER = Mod.loc("textures/screens/teammate/friendly_helicopter.png");
    public static final ResourceLocation FRIENDLY_MINE = Mod.loc("textures/screens/teammate/friendly_mine.png");

    @Override
    @ParametersAreNonnullByDefault
    public void render(GuiGraphics guiGraphics, DeltaTracker deltaTracker) {
        if (!DisplayConfig.VEHICLE_INFO.get()) return;

        Minecraft mc = Minecraft.getInstance();
        Player player = mc.player;
        var screenWidth = guiGraphics.guiWidth();
        var screenHeight = guiGraphics.guiHeight();
        var partialTick = deltaTracker.getGameTimeDeltaPartialTick(true);
        Camera camera = mc.gameRenderer.getMainCamera();
        Vec3 cameraPos = camera.getPosition();


        if (player == null) return;

        CuriosApi.getCuriosInventory(player).flatMap(c -> c.findFirstCurio(ModItems.IFF.get())).ifPresent(s -> {
            List<Entity> entities = SeekTool.getTeammate(player, player.level());
            for (var e : entities) {
                if (e != null && e != player && VectorUtil.canSee(e.position()) && e != player.getVehicle()) {
                    Entity team = e;
                    if (e.getVehicle() != null) {
                        team = e.getVehicle();
                    }

                    RenderSystem.disableDepthTest();
                    RenderSystem.depthMask(false);
                    RenderSystem.enableBlend();
                    RenderSystem.setShader(GameRenderer::getPositionTexShader);
                    RenderSystem.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);

                    if (checkNoClip(player, team, cameraPos)) {
                        RenderSystem.setShaderColor(1, 1, 1, 1);
                    } else {
                        RenderSystem.setShaderColor(1, 1, 1, 0.4f);
                    }

                    Vec3 pos = new Vec3(Mth.lerp(partialTick, team.xo, team.getX()), Mth.lerp(partialTick, team.yo + team.getBbHeight() / 2, team.getY() + team.getBbHeight() / 2), Mth.lerp(partialTick, team.zo, team.getZ()));
                    Vec3 point = VectorUtil.worldToScreen(pos);
                    float xf = (float) point.x;
                    float yf = (float) point.y;
                    ResourceLocation icon = getResourceLocation(team);

                    preciseBlit(guiGraphics, icon, Mth.clamp(xf - 6, 0, screenWidth - 12), Mth.clamp(yf - 6, 0, screenHeight - 12), 0, 0, 12, 12, 12, 12);
                }
            }
        });
    }

    private static ResourceLocation getResourceLocation(Entity team) {
        ResourceLocation icon = FRIENDLY_INDICATOR;

        // TODO 载具类型判断没生效？
        if ((team instanceof VehicleEntity vehicle && vehicle.getVehicleType() == VehicleEntity.VehicleType.BOAT) || team instanceof Boat) {
            icon = FRIENDLY_BOAT;
        } else if (team instanceof VehicleEntity vehicle) {
            if (vehicle.getVehicleType() == VehicleEntity.VehicleType.AIRPLANE) {
                icon = FRIENDLY_AIRCRAFT;
            }
            if (vehicle.getVehicleType() == VehicleEntity.VehicleType.HELICOPTER) {
                icon = FRIENDLY_HELICOPTER;
            }
            if (vehicle.getVehicleType() == VehicleEntity.VehicleType.APC) {
                icon = FRIENDLY_APC;
            }
            if (vehicle.getVehicleType() == VehicleEntity.VehicleType.CAR) {
                icon = FRIENDLY_CAR;
            }
            if (vehicle.getVehicleType() == VehicleEntity.VehicleType.AA) {
                icon = FRIENDLY_AA;
            }
            if (vehicle.getVehicleType() == VehicleEntity.VehicleType.TANK) {
                icon = FRIENDLY_TANK;
            }
            if (vehicle.getVehicleType() == VehicleEntity.VehicleType.ARTILLERY) {
                icon = FRIENDLY_ARTILLERY;
            }
            if (vehicle.getVehicleType() == VehicleEntity.VehicleType.DRONE) {
                icon = FRIENDLY_DRONE;
            }
            if (vehicle.getVehicleType() == VehicleEntity.VehicleType.DEFENSE) {
                icon = FRIENDLY_DEFENSE;
            }
        } else if (team instanceof MineEntity) {
            icon = FRIENDLY_MINE;
        }
        return icon;
    }

    public static boolean checkNoClip(Player player, Entity teammate, Vec3 pos) {
        return player.level().clip(new ClipContext(pos, teammate.position(),
                ClipContext.Block.COLLIDER, ClipContext.Fluid.ANY, player)).getType() != HitResult.Type.BLOCK;
    }

    public static double calculateAngle(Entity entityA, Camera camera) {
        Vec3 v1 = camera.getPosition().vectorTo(entityA.position());
        Vec3 v2 = new Vec3(camera.getLookVector());
        return VectorTool.calculateAngle(v1, v2);
    }
}

File: src/main/kotlin/com/atsuishio/superbwarfare/client/overlay/IFFOverlay.kt
package com.atsuishio.superbwarfare.client.overlay

import com.atsuishio.superbwarfare.Mod.Companion.loc
import com.atsuishio.superbwarfare.client.ClientSyncedEntityHandler
import com.atsuishio.superbwarfare.client.RenderHelper
import com.atsuishio.superbwarfare.config.client.DisplayConfig
import com.atsuishio.superbwarfare.data.vehicle.subdata.VehicleType
import com.atsuishio.superbwarfare.entity.projectile.MissileProjectile
import com.atsuishio.superbwarfare.entity.vehicle.base.VehicleEntity
import com.atsuishio.superbwarfare.init.ModItems
import com.atsuishio.superbwarfare.init.ModTags
import com.atsuishio.superbwarfare.network.message.receive.EntitySyncMessage
import com.atsuishio.superbwarfare.tools.*
import com.mojang.blaze3d.platform.GlStateManager
import com.mojang.blaze3d.systems.RenderSystem
import net.minecraft.client.Camera
import net.minecraft.client.renderer.GameRenderer
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.OwnableEntity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.entity.vehicle.Boat
import net.minecraft.world.level.ClipContext
import net.minecraft.world.phys.HitResult
import net.minecraft.world.phys.Vec2
import net.minecraft.world.phys.Vec3
import net.minecraftforge.api.distmarker.Dist
import net.minecraftforge.api.distmarker.OnlyIn
import net.minecraftforge.registries.ForgeRegistries
import top.theillusivec4.curios.api.CuriosApi

@OnlyIn(Dist.CLIENT)
object IFFOverlay : CommonOverlay("iff") {
    val FRIENDLY_INDICATOR = loc("textures/overlay/teammate/friendly_indicator.png")
    val FRIENDLY_AIRCRAFT = loc("textures/overlay/teammate/friendly_aircraft.png")
    val FRIENDLY_TANK = loc("textures/overlay/teammate/friendly_tank.png")
    val FRIENDLY_APC = loc("textures/overlay/teammate/friendly_apc.png")
    val FRIENDLY_AA = loc("textures/overlay/teammate/friendly_aa.png")
    val FRIENDLY_CAR = loc("textures/overlay/teammate/friendly_car.png")
    val FRIENDLY_ARTILLERY = loc("textures/overlay/teammate/friendly_artillery.png")
    val FRIENDLY_BOAT = loc("textures/overlay/teammate/friendly_boat.png")
    val FRIENDLY_DEFENSE = loc("textures/overlay/teammate/friendly_defense.png")
    val FRIENDLY_DRONE = loc("textures/overlay/teammate/friendly_drone.png")
    val FRIENDLY_HELICOPTER = loc("textures/overlay/teammate/friendly_helicopter.png")
    val FRIENDLY_MINE = loc("textures/overlay/teammate/friendly_mine.png")
    val FRIENDLY_MISSILE = loc("textures/overlay/teammate/friendly_missile.png")

    override fun shouldRender() = super.shouldRender() && DisplayConfig.VEHICLE_INFO.get()

    override fun RenderContext.render() {
        val level = player.level()

        val poseStack = guiGraphics.pose()
        poseStack.pushPose()

        val clientEntities = SeekTool.Builder(player)
            .friendly()
            .notPlayer()
            .build()

        for (e in clientEntities) {
            val friendlyList = arrayListOf<EntitySyncMessage.SyncedEntity>()
            val synced = EntitySyncMessage.SyncedEntity(
                e.id,
                ForgeRegistries.ENTITY_TYPES.getKey(e.type)!!,
                e.position(),
                e.deltaMovement,
                e.serializeNBT()
            )

            friendlyList.add(synced)
            ClientSyncedEntityHandler.sync(level.dimension().location(), friendlyList, true)
        }

        CuriosApi.getCuriosInventory(player).ifPresent { c ->
            c.findFirstCurio(ModItems.IFF.get()).ifPresent { _ ->
                val entities = ClientSyncedEntityHandler.getSyncedEntities(level)
                for (entity in entities) {
                    val e = level.getEntity(entity.id) ?: entity

                    if (e !== player && e.position().canBeSeen() && e !== player.vehicle) {
                        val teammate = e.vehicle ?: e

                        RenderSystem.disableDepthTest()
                        RenderSystem.depthMask(false)
                        RenderSystem.enableBlend()
                        RenderSystem.setShader { GameRenderer.getPositionTexShader() }
                        RenderSystem.blendFuncSeparate(
                            GlStateManager.SourceFactor.SRC_ALPHA,
                            GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA,
                            GlStateManager.SourceFactor.ONE,
                            GlStateManager.DestFactor.ZERO
                        )

                        if (checkNoClip(player, teammate, cameraPos)) {
                            RenderSystem.setShaderColor(1f, 1f, 1f, 1f)
                        } else {
                            RenderSystem.setShaderColor(1f, 1f, 1f, 0.4f)
                        }

                        val pos = VectorTool.lerpGetEntityBoundingBoxCenter(teammate, partialTick)
                        val point = pos.worldToScreen()
                        val xf = point.x.toFloat()
                        val yf = point.y.toFloat()
                        val icon = getResourceLocation(teammate)

                        RenderHelper.blit(
                            guiGraphics.pose(),
                            icon,
                            (xf - 6).coerceIn(0f, (screenWidth - 12).toFloat()),
                            (yf - 6).coerceIn(0f, (screenHeight - 12).toFloat()),
                            0f,
                            0f,
                            12f,
                            12f,
                            12f,
                            12f,
                            0x7FFFAD
                        )

                        if (Vec2(xf, yf).distanceToSqr(Vec2(screenWidth.toFloat() / 2.0f, screenHeight.toFloat() / 2.0f)) < 12) {
                            poseStack.pushPose()
                            poseStack.translate(xf, yf, 0f)
                            poseStack.scale(0.75f, 0.75f, 1f)
                            val str = "${e.displayName.string} [${FormatTool.format1D(pos.distanceTo(cameraPos))}m]"
                            guiGraphics.drawString(
                                mc.font,
                                str,
                                -mc.font.width(str) / 2,
                                10,
                                0x7FFFAD,
                                false
                            )
                            poseStack.popPose()
                        }

                        RenderSystem.setShaderColor(1f, 1f, 1f, 1f)
                    }
                }

                val players = ClientSyncedEntityHandler.getSyncedPlayerInfo()

                for (otherPlayers in players) {
                    if (otherPlayers.uuid != player.uuid) {
                        val localPlayer = EntityFindUtil.findPlayer(level, otherPlayers.uuid.toString())
                        var pos = otherPlayers.pos

                        if (localPlayer != null) {
                            pos = VectorTool.lerpGetEntityBoundingBoxCenter(localPlayer, partialTick)
                        }

                        if (pos.canBeSeen()) {
                            RenderSystem.disableDepthTest()
                            RenderSystem.depthMask(false)
                            RenderSystem.enableBlend()
                            RenderSystem.setShader { GameRenderer.getPositionTexShader() }
                            RenderSystem.blendFuncSeparate(
                                GlStateManager.SourceFactor.SRC_ALPHA,
                                GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA,
                                GlStateManager.SourceFactor.ONE,
                                GlStateManager.DestFactor.ZERO
                            )

                            if (checkNoClip(player, pos, cameraPos)) {
                                RenderSystem.setShaderColor(1f, 1f, 1f, 1f)
                            } else {
                                RenderSystem.setShaderColor(1f, 1f, 1f, 0.4f)
                            }

                            val point = pos.worldToScreen()
                            val xf = point.x.toFloat()
                            val yf = point.y.toFloat()

                            RenderHelper.blit(
                                guiGraphics.pose(),
                                FRIENDLY_INDICATOR,
                                (xf - 6).coerceIn(0f, (screenWidth - 12).toFloat()),
                                (yf - 6).coerceIn(0f, (screenHeight - 12).toFloat()),
                                0f,
                                0f,
                                12f,
                                12f,
                                12f,
                                12f,
                                0x7FFFAD
                            )

                            if (Vec2(xf, yf).distanceToSqr(Vec2(screenWidth.toFloat() / 2.0f, screenHeight.toFloat() / 2.0f)) < 12) {
                                poseStack.pushPose()
                                poseStack.translate(xf, yf, 0f)
                                poseStack.scale(0.75f, 0.75f, 1f)
                                val str = "${otherPlayers.name} [${FormatTool.format1D(pos.distanceTo(cameraPos))}m]"
                                guiGraphics.drawString(
                                    mc.font,
                                    str,
                                    -mc.font.width(str) / 2,
                                    10,
                                    0x7FFFAD,
                                    false
                                )
                                poseStack.popPose()
                            }

                            RenderSystem.setShaderColor(1f, 1f, 1f, 1f)
                        }
                    }
                }

                val hostileEntities = ClientSyncedEntityHandler.getSyncedHostileEntities(player.level())
                for (entity in hostileEntities) {
                    val e = level.getEntity(entity.id) ?: entity

                    if (e !== player && e.position().canBeSeen() && e !== player.vehicle) {
                        val enemy = e.vehicle ?: e

                        RenderSystem.disableDepthTest()
                        RenderSystem.depthMask(false)
                        RenderSystem.enableBlend()
                        RenderSystem.setShader { GameRenderer.getPositionTexShader() }
                        RenderSystem.blendFuncSeparate(
                            GlStateManager.SourceFactor.SRC_ALPHA,
                            GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA,
                            GlStateManager.SourceFactor.ONE,
                            GlStateManager.DestFactor.ZERO
                        )

                        if (checkNoClip(player, enemy, cameraPos)) {
                            RenderSystem.setShaderColor(1f, 1f, 1f, 1f)
                        } else {
                            RenderSystem.setShaderColor(1f, 1f, 1f, 0.4f)
                        }

                        val pos = VectorTool.lerpGetEntityBoundingBoxCenter(enemy, partialTick)
                        val point = pos.worldToScreen()
                        val xf = point.x.toFloat()
                        val yf = point.y.toFloat()
                        val icon = getResourceLocation(enemy)

                        var color = 0xFFBD7F

                        if (e is VehicleEntity && ((e.firstPassenger == null && e.team == null) || (e is OwnableEntity && e.owner == null))) {
                            color = -1
                        }

                        RenderHelper.blit(
                            guiGraphics.pose(),
                            icon,
                            (xf - 6).coerceIn(0f, (screenWidth - 12).toFloat()),
                            (yf - 6).coerceIn(0f, (screenHeight - 12).toFloat()),
                            0f,
                            0f,
                            12f,
                            12f,
                            12f,
                            12f,
                            color
                        )

                        if (Vec2(xf, yf).distanceToSqr(Vec2(screenWidth.toFloat() / 2.0f, screenHeight.toFloat() / 2.0f)) < 12) {
                            poseStack.pushPose()
                            poseStack.translate(xf, yf, 0f)
                            poseStack.scale(0.75f, 0.75f, 1f)
                            val str = "${e.displayName.string} [${FormatTool.format1D(pos.distanceTo(cameraPos))}m]"
                            guiGraphics.drawString(
                                mc.font,
                                str,
                                -mc.font.width(str) / 2,
                                10,
                                color,
                                false
                            )
                            poseStack.popPose()
                        }

                        RenderSystem.setShaderColor(1f, 1f, 1f, 1f)
                    }
                }
            }
        }

        poseStack.popPose()
    }

    private fun getResourceLocation(entity: Entity): ResourceLocation {
        return if (entity is Boat) {
            FRIENDLY_BOAT
        } else if (entity is VehicleEntity) {
            when (entity.vehicleType) {
                VehicleType.AIRPLANE -> FRIENDLY_AIRCRAFT
                VehicleType.HELICOPTER -> FRIENDLY_HELICOPTER
                VehicleType.APC -> FRIENDLY_APC
                VehicleType.CAR -> FRIENDLY_CAR
                VehicleType.AA -> FRIENDLY_AA
                VehicleType.TANK -> FRIENDLY_TANK
                VehicleType.ARTILLERY -> FRIENDLY_ARTILLERY
                VehicleType.DRONE -> FRIENDLY_DRONE
                VehicleType.BOAT -> FRIENDLY_BOAT
                VehicleType.DEFENSE -> FRIENDLY_DEFENSE
                else -> FRIENDLY_INDICATOR
            }
        } else if (entity.type.`is`(ModTags.EntityTypes.MINE)) {
            FRIENDLY_MINE
        } else if (entity is MissileProjectile) {
            FRIENDLY_MISSILE
        } else {
            FRIENDLY_INDICATOR
        }
    }

    fun checkNoClip(player: Player, teammate: Entity, pos: Vec3): Boolean {
        return player.level().clip(
            ClipContext(
                pos, teammate.position(),
                ClipContext.Block.COLLIDER, ClipContext.Fluid.ANY, player
            )
        ).type != HitResult.Type.BLOCK
    }

    fun checkNoClip(player: Player, targetPos: Vec3, pos: Vec3): Boolean {
        return player.level().clip(
            ClipContext(
                pos, targetPos,
                ClipContext.Block.COLLIDER, ClipContext.Fluid.ANY, player
            )
        ).type != HitResult.Type.BLOCK
    }

    fun calculateAngle(entityA: Entity, camera: Camera): Double {
        val v1 = camera.position.vectorTo(entityA.position())
        val v2 = Vec3(camera.lookVector)
        return v1.angleTo(v2)
    }
}
